x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/DateTimeArithmeticOperation.java

```diff
@@ -48,30 +51,85 @@ interface DatetimeArithmeticEvaluator extends TriFunction<Source, ExpressionEval
+            );
         }
         return super.resolveType();
     }
 
+    /**
+     * Override this to allow processing literals of type {@link EsqlDataTypes#DATE_PERIOD} when folding constants.
+     * Used in {@link DateTimeArithmeticOperation#fold}.
+     * @param left the left period
+     * @param right the right period
+     * @return the result of the evaluation
+     */
+    protected Period processDatePeriods(Period left, Period right) {
+        throw new UnsupportedOperationException("processing of time periods is unsupported for [" + op().symbol() + "]");
+    }
+
+    /**
+     * Override this to allow processing literals of type {@link EsqlDataTypes#TIME_DURATION} when folding constants.
+     * Used in {@link DateTimeArithmeticOperation#fold}.
+     * @param left the left duration
+     * @param right the right duration
+     * @return the result of the evaluation
+     */
+    protected Duration processTimeDurations(Duration left, Duration right) {
+        throw new UnsupportedOperationException("processing of date durations is unsupported for [" + op().symbol() + "]");
+    }
+
+    @Override
+    public final Object fold() {
+        DataType leftDataType = left().dataType();
+        DataType rightDataType = right().dataType();
+        if (leftDataType == DATE_PERIOD && rightDataType == DATE_PERIOD) {
+            // Both left and right expressions are temporal amounts; we can assume they are both foldable.
+            Period l = (Period) left().fold();
+            Period r = (Period) right().fold();
+            return processDatePeriods(l, r);
+        }
+        if (leftDataType == TIME_DURATION && rightDataType == TIME_DURATION) {
+            // Both left and right expressions are temporal amounts; we can assume they are both foldable.
+            Duration l = (Duration) left().fold();
+            Duration r = (Duration) right().fold();
+            return processTimeDurations(l, r);
+        }
+        return super.fold();
+    }
+
     @Override
     public Supplier<ExpressionEvaluator> toEvaluator(Function<Expression, Supplier<ExpressionEvaluator>> toEvaluator) {
         return dataType() == DataTypes.DATETIME
             ? () -> datetimes.apply(
                 source(),
-                toEvaluator.apply(argumentOfType(DataTypes::isDateTime)).get(),
-                (TemporalAmount) argumentOfType(EsqlDataTypes::isTemporalAmount).fold()
+                toEvaluator.apply(argumentWhich(DataTypes::isDateTime)).get(),
+                (TemporalAmount) argumentWhich(EsqlDataTypes::isTemporalAmount).fold()
             )
             : super.toEvaluator(toEvaluator);
     }
 
-    private Expression argumentOfType(Predicate<DataType> filter) {
+    private Expression argumentWhich(Predicate<DataType> filter) {
         return filter.test(left().dataType()) ? left() : filter.test(right().dataType()) ? right() : null;
     }
```

Comment:
There are two arguments - if one is DateTime, the other has to be TemporalAmount, otherwise flip them.
Easier then doing 4 method invocations of a method that has two nested method invocations internally.

There's already a type check in resolve() where the explicit types are determined.

Reply:
I'll simplify `toEvaluator` and `resolveType` to reduce the necessary number of method invocations.